### PR TITLE
Rework cancellation counting

### PR DESCRIPTION
**Pull request by Quinn-With-Two-Ns**
# Summary

The Go SDK notoriously tries to count commands and this has lead to all sorts of issues in the Go SDK with the counter getting out of sync. At some point we introduced `workflowExecutionIsCancelling`  to try to correct some counting issues when the workflow execution was cancelled. The reason workflow cancellation was messing with the counting is because when the workflow was cancelled and the cancel event was processed by the WFT state machine all the state-machines that were waiting on that context were generating their respective cancellation command. All these cancel commands were being generated before the Go SDK had processed the WFT started event for this WFT. This is problematic because the WFT started event sets the counter to `workflowTaskStartedEventID +2` so all those command went uncounted. The solution was to count these cancelled command separately and add them to the total. 

The Go SDK approach to cancellation also causes issues when the workflow is cancelled in the same WFT as other events that may resolve other state machines like `TimerCompleted` since the cancellation happens before those event are processed and state machines are resolved. 

Even with the 2+ years of work on this there are still issues with how this correction is implemented https://github.com/temporalio/sdk-go/issues/1176.

I explored changing how cancellation is processed so it is delayed until the WFT state machine processes the WFT started event. In the Go SDK the only safe place to generate commands is after a WFT started event is processed. Unfortunately that is a breaking change because if the workflow is cancelled in the same WFT as a event that resolves a state machine is sent the order they are processed matter.

Since we can't change how cancellation is processed I think the next best solution is base the correction on the unsent commands generated before the WFT started event was processed, rather then keep a separate counter.

This should not be a breaking change to any workflow that worked before this change, but I welcome any new tests to prove this

There is a more history here:
https://github.com/temporalio/sdk-go/pull/726 

There was also a test in this [issue](https://github.com/temporalio/sdk-go/issues/724) (that we couldn't add due to CPU usage?) that I ran as well

### CODE REVIEW

File: internal/internal_command_state_machine.go [hunk 1]
```diff
@@ -905,13 +874,21 @@ func (h *commandsHelper) setCurrentWorkflowTaskStartedEventID(workflowTaskStarte
...
-	h.nextCommandEventIDResetCounter++
+	// command. It is possible, notably during workflow cancellation, that commands are generated before the workflow
+	// task started event is processed. In this case we need to adjust the nextCommandEventID to account for these unsent
+	// commands.git
+	var uncountedCommands int64
+	for curr := h.orderedCommands.Front(); curr != nil; {
+		d := curr.Value.(commandStateMachine)
+		command := d.getCommand()
+		if command != nil {
+			fmt.Printf("command: %v \n", command)
```
Comment (Quinn-With-Two-Ns) [inline internal/internal_command_state_machine.go:885]: Ah debug print!

**Review by Sushisource — APPROVED** [review]
This makes sense to me and is definitely cleaner than before

File: test/workflow_test.go [hunk 2]
```diff
@@ -1571,6 +1571,59 @@ func (w *Workflows) CronWorkflow(ctx workflow.Context) (int, error) {
...
+	defer func() {
+		if !errors.Is(ctx.Err(), workflow.ErrCanceled) {
+			return
+		}
+		// When the Workflow is canceled, it has to get a new disconnected context to execute any Activities
+		newCtx, _ := workflow.NewDisconnectedContext(ctx)
+		err = workflow.ExecuteActivity(newCtx, activities.EmptyActivity).Get(newCtx, nil)
+	}()
+
+	//Create selector
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply test/workflow_test.go:1592]: yep i'll remove, this whole test was taken from the bug report .  Must have missed this comment

File: internal/internal_command_state_machine.go [hunk 3]
```diff
@@ -905,13 +874,21 @@ func (h *commandsHelper) setCurrentWorkflowTaskStartedEventID(workflowTaskStarte
...
-	// We must change the counter here so that others who mutate
-	// commandsCancelledDuringWFCancellation know it has since been reset
-	h.nextCommandEventIDResetCounter++
+	// command. It is possible, notably during workflow cancellation, that commands are generated before the workflow
+	// task started event is processed. In this case we need to adjust the nextCommandEventID to account for these unsent
+	// commands.git
+	var uncountedCommands int64
+	for curr := h.orderedCommands.Front(); curr != nil; {
+		d := curr.Value.(commandStateMachine)
+		command := d.getCommand()
```
Comment (cretz) [inline internal/internal_command_state_machine.go:883]: I wish this call were `createCommand()`. It was unclear to me what was happening here until I understood that's what this is.

**Review by cretz — COMMENTED** [review]
If all the tests pass, I support any fixing of counters in any way. Will hold off approval until debug print is removed.

**Comment by Quinn-With-Two-Ns** [comment]
@cretz done

**Review by cretz — APPROVED** [review]

**Pull request merged**